sdk/java/pulumi/src/main/java/com/pulumi/experimental/automation/serialization/internal/ProjectSettingsYamlSerializer.java

```diff
@@ -0,0 +1,386 @@
+            super(options);
+            this.addClassTag(ProjectSettings.class, Tag.MAP);
+            this.addClassTag(ProjectRuntimeName.class, Tag.STR);
+            this.getPropertyUtils().setBeanAccess(BeanAccess.FIELD);
+            this.getPropertyUtils().setSkipMissingProperties(true);
+
+            representers.put(ProjectRuntimeName.class,
+                    data -> represent(((ProjectRuntimeName) data).getValue()));
+        }
+
+        @Override
+        protected NodeTuple representJavaBeanProperty(Object javaBean, Property property,
+                Object propertyValue, Tag customTag) {
+            // Skip null values
+            if (propertyValue == null) {
+                return null;
+            }
+
+            // Skip empty collections
+            if (propertyValue instanceof Collection && ((Collection<?>) propertyValue).isEmpty()) {
+                return null;
+            }
+
+            // Skip empty maps
+            if (propertyValue instanceof Map && ((Map<?, ?>) propertyValue).isEmpty()) {
+                return null;
+            }
+
+            // Skip Boolean values that are false and annotated with @SkipIfFalse
+            if (propertyValue instanceof Boolean && !(Boolean) propertyValue) {
+                try {
+                    var field = javaBean.getClass().getDeclaredField(property.getName());
+                    if (field.isAnnotationPresent(SkipIfFalse.class)) {
+                        return null;
+                    }
+                } catch (NoSuchFieldException e) {
+                    // ignore
+                }
+            }
+
+            // Serialize properties with a trailing underscore as the name without the
+            // underscore.
+            // For example "default_" will be serialized as "default".
+            var name = property.getName();
+            if (name.endsWith("_")) {
+                var result = super.representJavaBeanProperty(javaBean, property, propertyValue, customTag);
+                return new NodeTuple(
+                        representScalar(Tag.STR, name.substring(0, name.length() - 1)),
+                        result.getValueNode());
+            }
+
+            // Special handling of ProjectRuntime
+            if (propertyValue instanceof ProjectRuntime) {
+                var runtime = (ProjectRuntime) propertyValue;
+                if (runtime.getOptions() == null) {
+                    // If options is null, represent runtime as just the name
+                    return new NodeTuple(
+                            representScalar(Tag.STR, property.getName()),
+                            representScalar(Tag.STR, runtime.getName().name().toLowerCase()));
+                }
+            }
+
+            return super.representJavaBeanProperty(javaBean, property, propertyValue, customTag);
+        }
+    }
+
+    public static class ProjectSettingsModel implements Model<ProjectSettings> {
```

Comment:
I was trying to make the (de)serialization preserve formatting and comments when round tripping, but ran out of time. I was also seeing if I could get snakeYAML to deserialize to the builders themselves, but couldn't get it working, so went with these internal model classes that the YAML gets deserialized to, and then convert those to the actual public classes. I'm not a fan of all this boilerplate, but I guess it's no worse than .NET: https://github.com/pulumi/pulumi-dotnet/tree/main/sdk/Pulumi.Automation/Serialization

I briefly looked into switching to Jackson for YAML (and JSON) parsing, but it looks like there's not as much control over preserving comments there -- which I'd love to get back to at some point. So I've kept snakeYAML for now.